Make trigger log available in django

tilboerner opened the pull request:
To diagnose and repair Heroku Connect sync errors in a Django environment, the first step is to make Heroku Connect's trigger log available there. This pull request:

- makes the `_trigger_log` and `_trigger_log_archive` tables accessible as (unmanaged) models `TriggerLog` and `TriggerLogArchive`;
- introduces a (managed) `TriggerLogPermanent` model that allows retaining trigger log data beyond its lifetime in the Heroku Connect tables;
- gives access to the PostgreSQL stored procedures that Heroku Connect provides to repair sync errors;
- includes an admin for trigger log models, so trigger log history can be explored.

In addition, to facilitate matching trigger log entries with their respective connected model instances, `HerokuConnectModelBase` gets amended so connected models can be looked up by their table names, and vice versa.

~~The admin code is still lacking test coverage.~~

tilboerner commented:
@codingjoe I added [error_handling.rst](docs/error_handling.rst) to the docs, with a short introduction and an `:automodule:` that captures the TriggerLog* doc comments.

codingjoe commented on heroku_connect/admin.py:
```diff
@@ -0,0 +1,139 @@
+    def state_label(self, log):
+        state = log.state
+        mod = {
+            TriggerLogState.SUCCESS: 'success',
```
i18n the literals should be gettext translatables

codingjoe replied:
I would also recommend to put the map somewhere on module or class level to make it accessable/reusable in the future.

codingjoe commented on heroku_connect/admin.py:
```diff
@@ -0,0 +1,139 @@
+        action = log.action
+        return format_html('<span class="label label-default">{action}</span>', action=action)
+    action_label.allow_tags = True
+    action_label.short_description = 'Action'
```
i18n

codingjoe commented on heroku_connect/admin.py:
```diff
@@ -0,0 +1,139 @@
+        return format_html('<span class="label label-default">{action}</span>', action=action)
+    action_label.allow_tags = True
+    action_label.short_description = 'Action'
+    action_label.admin_order_field = 'action'
```
i18n

codingjoe commented on heroku_connect/admin.py:
```diff
@@ -0,0 +1,139 @@
+    appended, for example ``_changelist``.
+    """
+    model = model_or_instance if isinstance(model_or_instance, type) else type(model_or_instance)
+    content_type = ContentType.objects.get_for_model(model)
```
why do you need the content type to build an admin url? You can get the app_label and model name from the model or an instance directly. `model=cls.__name__` `app_label=cls._meta.app_label`

codingjoe commented on heroku_connect/db/models/base.py:
```diff
@@ -4,7 +4,27 @@
+__all__ = ('HerokuConnectModel', 'registry')
+
+
+class HerokuConnectModelRegistry:
```
We already have a `utils.get_heroku_connect_models` wich uses Django's buildin model registry. No need to add yet another one ;)

tilboerner replied:
I thought it would be nicer to not always iterate over every connected model and transform its `db_table` to match a `table_name` from the trigger log. But I guess you might call the registry a premature optimisation; it definitely shares some of the functionality of `get_heroku_connect_models`.

codingjoe commented on heroku_connect/models.py:
```diff
@@ -0,0 +1,342 @@
+
+    @classmethod
+    def choices(cls):
+        return tuple((getattr(cls, name), name) for name in dir(cls) if name.isupper())
```
Don't you think this is utterly complex for:
```
('INSERT', 'INSERT'), ('UPDATE', 'UPDATE'), ('DELETE','DELETE')
```
?

codingjoe replied:
If you want to use the class as django choices, I would rather recommend to build a helper for that.
```python
def cls_to_choices(cls):
    return cls.__dict__.items()
```

tilboerner replied:
Ideally I'd like Django to have a choices-enum class! ;P My idea was to:

- have the choices adapt automatically when the class attributes change;
- not duplicate the much bigger list of states on the other class;
- make it clear on the class itself that the attributes will be used as choices;
- avoid using a dependency  for it.

So I'd prefer to leaving the method on the class instead of using a helper from somewhere else. I could introduce a super class to avoid duplicating the method code. Using `__dict__` is fine, but it still needs to at least filter out names starting with `_` or maybe only `__`.

tilboerner replied:
I'll just ause a `dict` instead.

codingjoe commented on heroku_connect/models.py:
```diff
@@ -0,0 +1,342 @@
+    """An archived entry that was copied back into as NEW into the current log."""
+
+    @classmethod
+    def choices(cls):
```
In any case, this would be duplication ;)

codingjoe commented on heroku_connect/models.py:
```diff
@@ -0,0 +1,342 @@
+    action = models.CharField(max_length=7, editable=False, choices=TriggerLogAction.choices())
+    sf_message = models.TextField(editable=False, null=True, blank=True)
+
+    # TODO: these are more useful as HStoreFields (if 'django.contrib.postgres' in INSTALLED_APPS)
```
Just document it as a dependency. It's not like you can you any other database with Heroku Connect ;)
Please make sure to add extension to the initial migration:
https://docs.djangoproject.com/en/2.0/ref/contrib/postgres/operations/#hstoreextension
It isn't added automatically :/

tilboerner replied:
@codingjoe Almost done addressing the feedback bit by bit, just getting around to this...

> Please make sure to add [hstore] extension to the initial migration:

This will make the migration fail if the database user lacks the appropriate privileges. Maybe it's better to add a check which looks at [`pg_available_extensions`](https://www.postgresql.org/docs/10/static/view-pg-available-extensions.html) and prints instructions to add the extension?

codingjoe commented on heroku_connect/models.py:
```diff
@@ -0,0 +1,342 @@
+    state = models.CharField(max_length=8, null=False, blank=False,
+                             choices=TriggerLogState.choices())
+
+    objects = models.Manager.from_queryset(TriggerLogQuerySet)()
```
`Queryset.as_manager()` ;)

codingjoe commented on heroku_connect/models.py:
```diff
@@ -0,0 +1,342 @@
+
+    @classmethod
+    def capture_insert_from_model(cls, table_name, record_id, *, exclude_fields=()):
+        """Create a fresh insert record from the current model state in the database.
```
I prefer if multi line docstrings start with a newline. PEP257 doesn't define that, but it just looks nicer, since all lines start with the same base indentation. Makes it easer for the eyes to read it.

codingjoe commented on heroku_connect/models.py:
```diff
@@ -0,0 +1,342 @@
+            ) AS id
+            FROM "{schema}"."{table_name}"
+            WHERE id = %(record_id)s
+        """.format(
```
If you want to write and your own sql I recommend using psycopg2's native formating and escaping functionality, see http://initd.org/psycopg/docs/sql.html

codingjoe commented on heroku_connect/models.py:
```diff
@@ -0,0 +1,342 @@
+            exclude_cols=', '.join("'{}'".format(col) for col in exclude_cols)
+        )
+        params = {'record_id': record_id}
+        return list(TriggerLog.objects.raw(sql, params))
```
memory: Are you sure you want to return a list over not an iterable?

codingjoe commented on heroku_connect/models.py:
```diff
@@ -0,0 +1,342 @@
+            include_cols=', '.join("'{}'".format(col) for col in include_cols),
+        )
+        params = {'record_id': record_id}
+        return list(TriggerLog.objects.raw(sql, params))
```
memory

codingjoe commented on heroku_connect/models.py:
```diff
@@ -0,0 +1,342 @@
+            SELECT "{schema}".hc_capture_update_from_row(
+              hstore({schema}.{table_name}.*),
+              '{table_name}',
+              ARRAY[{include_cols}]::text[]
```
type casts can be expensive expensive, is there any way to store these values in the desired types?

codingjoe commented on heroku_connect/models.py:
```diff
@@ -0,0 +1,342 @@
+        verbose_name = 'Archived Trigger Log'
+
+
+class TriggerLogPermanent(TriggerLogAbstract):
```
I don't know about this one. I don't think Postgres is a good place for logging data to begin with and even less as an archive. I like the idea of retaining them, but only via `syslog`. Users can decide than if you want to persist these logs. EG. we use S3 to persist logs over 30 days old.

codingjoe commented on tests/conftest.py:
```diff
@@ -0,0 +1,130 @@
+        for cls in model_classes:
+            schema.create_model(cls)
+    yield
+    # let django TestCase transaction rollback take care of cleaning this up again
```
?????

codingjoe commented on tests/conftest.py:
```diff
@@ -0,0 +1,93 @@
+    kwargs.setdefault('record_id', model.id)
+    kwargs.setdefault('created_at', timezone.now())
+    log = make_trigger_log(is_archived=is_archived, **kwargs)
+    log.save()
```
I don't like already saved model fixtures. They hide IO time. I want to see those expensive DB call in the test itself ;)

codingjoe commented on heroku_connect/models.py:
```diff
@@ -0,0 +1,274 @@
+    # are for technical folks and not intended to be user-facing.
+
+    # read-only fields
+    id = models.BigIntegerField(primary_key=True, editable=False)
```
Isn't this a serial, meaning it has a default `nextval(sequence)`.
I believe those are positive integers, but I would use the AutoField, which is the default. So I would just drop the line.

tilboerner replied:
I see this as explicit documentation of the unmanaged table: yes, this column exists, and it's some form of big integer PK. We're not managing it, so it might have autofield behavior, or it might not. I could make it an `AutoField` for easier testing, but I'd really like to have it there explicitly.

codingjoe replied:
Please use the `AutoField` than. I check it does have `nextval('salesforce._trigger_log_id_seq'::regclass)` as a default, so is a postgres `serial` not `int`.

codingjoe replied:
We don't have it for the BaseModel ether. If you want to overwrite the default go for it, I don't mind ether way.

tilboerner replied:
Github is showing old code here, because I already changed the line. It's currently a `BigAutoField`.

codingjoe commented on heroku_connect/models.py:
```diff
@@ -0,0 +1,274 @@
+
+    # read-only fields
+    id = models.BigIntegerField(primary_key=True, editable=False)
+    created_at = models.DateTimeField(editable=False, null=True)
```
None of the fields are editable, are they?
I don't know. I would just drop it and make sure to have them as read_only in the admin.

tilboerner replied:
Same as above: it's documentation that the field shouldn't be touched, and a hedge against somebody creating a `ModelForm` and messing things up. I think it's better to make that point right where stuff is defined, not in the admin.

codingjoe commented on heroku_connect/models.py:
```diff
@@ -0,0 +1,274 @@
+    state = models.CharField(max_length=8, null=False, blank=False,
+                             choices=TRIGGER_LOG_STATE_CHOICES)
+
+    objects = models.Manager.from_queryset(TriggerLogQuerySet)()
```
`TriggerLogQuerySet.as_manager()`

codingjoe commented on heroku_connect/models.py:
```diff
@@ -0,0 +1,274 @@
+    _old = models.TextField(editable=False, null=True, blank=True, db_column='old')
+
+    # editable fields
+    state = models.CharField(max_length=8, null=False, blank=False,
```
ok, except this one... still ;)

codingjoe commented on heroku_connect/models.py:
```diff
@@ -0,0 +1,274 @@
+            model_cls = get_connected_model_for_table_name(table_name)
+            exclude_cols = cls._fieldnames_to_colnames(model_cls, exclude_fields)
+
+        raw_query = sql.SQL("""
```
👏

codingjoe replied:
Please show this to the others on Friday.

tilboerner replied:
You mean, how to be safe with `sql.SQL`? Sure!

codingjoe replied:
yes

codingjoe commented:
btw, even unmanaged models need migrations ;)

tilboerner commented:
@codingjoe I have the migration queued up for a commit, I was just deciding whether to create the hstore extension in it or not. (There are some HStore fields in the models.) Did you see my related comment? Basically, if we create the extension in the migration, it'll need to be run by a DB superuser.

The alternative I see is to have a check for the presence of the extension with the advice to create the extension manually. What do you think?

tilboerner commented:
@codingjoe I addressed your comments and rebased off the current master. Rebasing unfortunately grouped all commits together here in the UI; my new commits are the ones after your `rm double new line`.

codingjoe commented on tests/test_models.py:
```diff
@@ -0,0 +1,107 @@
+class TestTriggerLog:
+
+    @pytest.fixture()
+    def _hstore_extension(self):
```
This can go now, can't it?

codingjoe commented:
@tilboerner let's test it tomorrow together with @samifahed. I think this is the last step remaining.
